# Day 17: Filtering and Sorting DataFrames

**Dataset:** `Superstore_Dataset.csv`

This notebook filters rows using conditions and sorts DataFrame values in ascending and
descending order, using real values pulled directly from the Superstore dataset.

## 1. Import Libraries and Load the Dataset

In [1]:
import pandas as pd

# Load the dataset (encoding='latin1' avoids a UnicodeDecodeError on this file)
df = pd.read_csv("../datasets/Superstore_Dataset.csv", encoding="latin1")

# Preview the data
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [2]:
print(df.shape)
df.info()

(9994, 21)
<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity

## 2. Filtering Examples

Filtering uses a **boolean condition** inside `df[...]` (or `df.loc[...]`) to keep only the
rows where that condition is `True`.

**Example 1 — Orders with Sales greater than 500**

In [3]:
high_sales = df[df["Sales"] > 500]
print(len(high_sales), "orders with Sales > 500")
high_sales[["Order ID", "Sales", "Category"]].head()

1162 orders with Sales > 500


,Order ID,Sales,Category
1,CA-2016-152156,731.9400,Furniture
3,US-2015-108966,957.5775,Furniture
7,CA-2014-115812,907.1520,Technology
10,CA-2014-115812,1706.1840,Furniture
11,CA-2014-115812,911.4240,Technology


**Example 2 — Orders in the "Furniture" category**

In [4]:
furniture_orders = df[df["Category"] == "Furniture"]
print(len(furniture_orders), "Furniture orders")
furniture_orders[["Order ID", "Category", "Sub-Category", "Sales"]].head()

2121 Furniture orders


,Order ID,Category,Sub-Category,Sales
0,CA-2016-152156,Furniture,Bookcases,261.9600
1,CA-2016-152156,Furniture,Chairs,731.9400
3,US-2015-108966,Furniture,Tables,957.5775
5,CA-2014-115812,Furniture,Furnishings,48.8600
10,CA-2014-115812,Furniture,Tables,1706.1840


**Example 3 — Loss-making orders (Profit less than 0)**

In [5]:
loss_making = df[df["Profit"] < 0]
print(len(loss_making), "loss-making orders")
loss_making[["Order ID", "Category", "Profit"]].head()

1871 loss-making orders


,Order ID,Category,Profit
3,US-2015-108966,Furniture,-383.0310
14,US-2015-118983,Office Supplies,-123.8580
15,US-2015-118983,Office Supplies,-3.8160
23,US-2017-156909,Furniture,-1.0196
27,US-2015-150630,Furniture,-1665.0522


**Example 4 — Combining multiple conditions with `&` (AND): loss-making Technology orders**

In [6]:
loss_making_tech = df[(df["Category"] == "Technology") & (df["Profit"] < 0)]
print(len(loss_making_tech), "loss-making Technology orders")
loss_making_tech[["Order ID", "Sub-Category", "Sales", "Profit"]].head()

271 loss-making Technology orders


,Order ID,Sub-Category,Sales,Profit
100,CA-2016-158568,Accessories,95.976,-10.7973
103,US-2015-156867,Accessories,238.896,-26.8758
130,US-2017-164147,Phones,59.970,-11.9940
165,CA-2014-139892,Machines,8159.952,-1359.9920
214,CA-2015-146262,Phones,62.982,-14.6958


**Example 5 — Combining multiple conditions with `|` (OR): orders from the West or East region**

In [7]:
west_or_east = df[(df["Region"] == "West") | (df["Region"] == "East")]
print(len(west_or_east), "orders from West or East")
west_or_east[["Order ID", "Region", "Sales"]].head()

6051 orders from West or East


,Order ID,Region,Sales
2,CA-2016-138688,West,14.620
5,CA-2014-115812,West,48.860
6,CA-2014-115812,West,7.280
7,CA-2014-115812,West,907.152
8,CA-2014-115812,West,18.504


**Example 6 — Filtering with `.isin()`: orders for specific sub-categories**

In [8]:
selected_subcategories = df[df["Sub-Category"].isin(["Phones", "Chairs", "Tables"])]
print(len(selected_subcategories), "orders in Phones, Chairs, or Tables")
selected_subcategories[["Order ID", "Sub-Category", "Sales"]].head()

1825 orders in Phones, Chairs, or Tables


,Order ID,Sub-Category,Sales
1,CA-2016-152156,Chairs,731.9400
3,US-2015-108966,Tables,957.5775
7,CA-2014-115812,Phones,907.1520
10,CA-2014-115812,Tables,1706.1840
11,CA-2014-115812,Phones,911.4240


## 3. Sorting Examples

`sort_values()` reorders a DataFrame by one or more columns, in ascending order by default.

**Example 1 — Sort by Sales, descending (highest sales first)**

In [9]:
df.sort_values("Sales", ascending=False)[["Order ID", "Sales"]].head(10)

,Order ID,Sales
2697,CA-2014-145317,22638.480
6826,CA-2016-118689,17499.950
8153,CA-2017-140151,13999.960
2623,CA-2017-127180,11199.968
4190,CA-2017-166709,10499.970
9039,CA-2016-117121,9892.740
4098,CA-2014-116904,9449.950
4277,US-2016-107440,9099.930
8488,CA-2016-158841,8749.950
6425,CA-2016-143714,8399.976


**Example 2 — Sort by Profit, ascending (biggest losses first)**

In [10]:
df.sort_values("Profit", ascending=True)[["Order ID", "Category", "Profit"]].head(10)

,Order ID,Category,Profit
7772,CA-2016-108196,Technology,-6599.9780
683,US-2017-168116,Technology,-3839.9904
9774,CA-2014-169019,Office Supplies,-3701.8928
3011,CA-2017-134845,Technology,-3399.9800
4991,US-2017-122714,Office Supplies,-2929.4845
3151,CA-2015-147830,Technology,-2639.9912
5310,CA-2017-131254,Office Supplies,-2287.7820
9639,CA-2015-116638,Furniture,-1862.3124
1199,CA-2016-130946,Office Supplies,-1850.9464
2697,CA-2014-145317,Technology,-1811.0784


**Example 3 — Sort by multiple columns: Category (A-Z), then Sales (highest first) within each category**

In [11]:
df.sort_values(["Category", "Sales"], ascending=[True, False])[["Category", "Sub-Category", "Sales"]].head(15)

,Category,Sub-Category,Sales
7243,Furniture,Chairs,4416.174
9741,Furniture,Bookcases,4404.900
9639,Furniture,Tables,4297.644
5917,Furniture,Bookcases,4228.704
6535,Furniture,Bookcases,4007.840
9425,Furniture,Chairs,3785.292
1246,Furniture,Tables,3610.848
5170,Furniture,Chairs,3504.900
9649,Furniture,Bookcases,3406.664
9857,Furniture,Tables,3393.680


**Example 4 — Sort the filtered loss-making orders by how large the loss is**

In [12]:
loss_making.sort_values("Profit", ascending=True)[["Order ID", "Category", "Profit"]].head(10)

,Order ID,Category,Profit
7772,CA-2016-108196,Technology,-6599.9780
683,US-2017-168116,Technology,-3839.9904
9774,CA-2014-169019,Office Supplies,-3701.8928
3011,CA-2017-134845,Technology,-3399.9800
4991,US-2017-122714,Office Supplies,-2929.4845
3151,CA-2015-147830,Technology,-2639.9912
5310,CA-2017-131254,Office Supplies,-2287.7820
9639,CA-2015-116638,Furniture,-1862.3124
1199,CA-2016-130946,Office Supplies,-1850.9464
2697,CA-2014-145317,Technology,-1811.0784


## 4. A Short Analysis of Filtered Records

Taking a closer look at the loss-making orders filtered above.

In [13]:
print("Total loss-making orders:", len(loss_making))
print("Total profit lost:", round(loss_making["Profit"].sum(), 2))
print()
print("Loss-making orders by category:")
print(loss_making["Category"].value_counts())
print()
print("Average loss per order, by category:")
print(loss_making.groupby("Category")["Profit"].mean().round(2))

Total loss-making orders: 1871
Total profit lost: -156131.29

Loss-making orders by category:
Category
Office Supplies    886
Furniture          714
Technology         271
Name: count, dtype: int64

Average loss per order, by category:
Category
Furniture          -85.34
Office Supplies    -63.90
Technology        -142.36
Name: Profit, dtype: float64


**Interpretation:** Out of 9,994 total orders, **1,871 (about 19%)** lost money, totaling
roughly **-$156,131** in lost profit. **Office Supplies** has the most loss-making orders
(886), followed by **Furniture** (714) and **Technology** (271) — but that's partly because
Office Supplies also has the most orders overall. Looking at the *average* loss per order by
category would be the next useful step to see which category loses the most money per
transaction, rather than just which has the most loss-making orders in total.

## Outcome

This notebook practiced filtering DataFrame rows using boolean conditions on the
`Superstore_Dataset.csv` file — single conditions (`Sales > 500`, `Category == "Furniture"`,
`Profit < 0`), combined conditions using `&` (AND) and `|` (OR), and membership filtering with
`.isin()`. I used `sort_values()` to sort by a single column in both descending and ascending
order, and to sort by multiple columns at once with independent ascending/descending settings
for each. Filtering down to the loss-making orders and analyzing them with `.value_counts()`
and `.groupby()` showed how filtering isn't just about narrowing down rows — it's a starting
point for asking further questions about the data, like which category loses the most money
and how that compares once order volume is accounted for.